algorithms/src/std_algorithms/Kokkos_NonModifyingSequenceOperations.hpp

```diff
@@ -0,0 +1,2100 @@
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_NON_MODIFYING_SEQUENCE_OPERATIONS_HPP
+#define KOKKOS_NON_MODIFYING_SEQUENCE_OPERATIONS_HPP
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_BeginEnd.hpp"
+#include "Kokkos_Constraints.hpp"
+#include "Kokkos_MinMaxOperations.hpp"
+#include "Kokkos_ModifyingOperations.hpp"
+#include "Kokkos_HelperPredicates.hpp"
+#include "Kokkos_RandomAccessIterator.hpp"
+#include <string>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+// ------------------------------------------
+//
+// functors
+//
+// ------------------------------------------
+
+template <bool is_find_if, class IndexType, class IteratorType,
+          class ReducerType, class PredicateType>
+struct StdFindIfOrNotFunctor {
+  using red_value_type = typename ReducerType::value_type;
+
+  IteratorType m_first;
+  ReducerType m_reducer;
+  PredicateType m_p;
+
+  KOKKOS_FUNCTION
+  void operator()(const IndexType i, red_value_type& red_value) const {
+    const auto& my_value = m_first[i];
+
+    // if doing find_if, look for when predicate is true
+    // if doing find_if_not, look for when predicate is false
+    const bool found_condition = is_find_if ? m_p(my_value) : !m_p(my_value);
+
+    auto rv =
+        found_condition
+            ? red_value_type{i}
+            : red_value_type{::Kokkos::reduction_identity<IndexType>::min()};
+
+    m_reducer.join(red_value, rv);
+  }
+
+  KOKKOS_FUNCTION
+  StdFindIfOrNotFunctor(IteratorType first, ReducerType reducer,
+                        PredicateType p)
+      : m_first(first), m_reducer(std::move(reducer)), m_p(std::move(p)) {}
+};
+
+template <class IteratorType, class UnaryFunctorType>
+struct StdForEachFunctor {
+  using index_type = typename IteratorType::difference_type;
+  IteratorType m_first;
+  UnaryFunctorType m_functor;
+
+  KOKKOS_FUNCTION
+  void operator()(index_type i) const { m_functor(m_first[i]); }
+
+  KOKKOS_FUNCTION
+  StdForEachFunctor(IteratorType _first, UnaryFunctorType _functor)
+      : m_first(_first), m_functor(std::move(_functor)) {}
+};
```

Comment:
Global comment:

Do we also want to add deduction guides (the implicitly generated one might cause a warning :-() as well as a function to deduce the arguments, as in:

```
template<class IteratorType, class UnaryFunctorType>
StdForEachFunctor(IteratorType, UnaryFunctorType) -> StdForEachFunctor<IteratorType, UnaryFunctorType>;

template<class IteratorType, class UnaryFunctorType>
KOKKOS_FUNCTION
StdForEachFunctor<IteratorType, UnaryFunctorType>
make_StdForEachFunctor(IteratorType _first, UnaryFunctorType _functor) {
    return StdForEachFunctor<IteratorType, UnaryFunctorType>(std::move(_first), std::move(_functor));
}
```

Reply:
Is deduction guide a c++17 thing? so should we do that now? I thought by end of this year Kokkos requires at least c++14 . am i missing something?

Reply:
It currently requires C++14, and Kokkos 4 will require C++17.

Reply:
yes, makes sense. So you think we should add deduction guides now or only after we adopt c++17?